File: tutorpaes/backend/app/schemas/quiz.py
from pydantic import BaseModel
from typing import List, Optional, Dict, Literal


class ChoiceOut(BaseModel):
    id: int
    label: str
    text: str


class QuestionOut(BaseModel):
    kind: Literal["question"] = "question"
    question_id: int
    prompt: str
    topic: str
    reading_text: Optional[str] = None
    choices: List[ChoiceOut]


class AnswerIn(BaseModel):
    # `user_id` se deriva del JWT (get_current_user). Se mantiene opcional por compatibilidad hacia atrás.
    user_id: Optional[int] = None
    subject_code: str
    topic_code: str
    question_id: int
    selected_choice_id: int


class AnswerOut(BaseModel):
    attempt_id: int
    feedback_id: int
    is_correct: bool
    feedback_text: str
    is_attempt_finished: bool
    ai_payload: Dict = {}


class TopicCompletedOut(BaseModel):
    kind: Literal["topic_completed"] = "topic_completed"
    message: str
    attempt_id: int
    status: str
    total_questions: int
    correct_count: int
    score_percentage: int
    score_paes: int
    score: int


class AIFeedbackOut(BaseModel):
    explanation: str
    is_correct: bool
    source: str
    correct_choice_id: Optional[int] = None
    correct_choice_label: Optional[str] = None


class AttemptFeedbackDetailOut(BaseModel):
    question_id: int
    prompt: str
    reading_text: Optional[str] = None
    selected_choice_id: Optional[int]
    selected_choice_text: Optional[str]
    correct_choice_id: Optional[int]
    correct_choice_text: Optional[str]
    is_correct: bool
    ai_explanation: Optional[str] = None


class AttemptResultOut(BaseModel):
    attempt_id: int
    exam_id: int
    subject_id: int
    topic_id: Optional[int]
    status: str
    score: Optional[int]
    total_questions: int
    correct_count: int
    started_at: str
    completed_at: Optional[str] = None
    answers_detail: List[AttemptFeedbackDetailOut]

File: tutorpaes/frontend/src/types/quiz.ts
/**
 * Quiz Types - Aligned with Backend Pydantic Schemas
 * 
 * These types MUST match exactly with:
 * - backend/app/schemas/quiz.py
 */

// ============================================================================
// BACKEND RESPONSE TYPES (from backend/app/schemas/quiz.py)
// ============================================================================

export interface BackendQuestionChoice {
  id: number;        // Backend uses "id", NOT "choice_id"
  label: string;     // "A", "B", "C", "D"
  text: string;
}

export interface BackendQuestionOut {
  kind: "question";
  question_id: number;
  prompt: string;    // Backend uses "prompt", NOT "text"
  topic: string;     // Topic code (e.g., "ALG", "GEO")
  reading_text?: string | null;
  choices: BackendQuestionChoice[];
  correct_choice_id: number; // ADDED: For correct proactive logic
}

export interface BackendTopicCompletedOut {
  kind: "topic_completed";
  message: string;
  attempt_id: number;
  status: string;
  total_questions: number;
  correct_count: number;
  score_percentage: number;
  score_paes: number;
  score: number;
}

export type NextQuestionResponse = BackendQuestionOut | BackendTopicCompletedOut;

// ============================================================================
// ANSWER SUBMISSION
// ============================================================================

export interface BackendAnswerIn {
  subject_code: string;     // REQUIRED
  topic_code: string;       // REQUIRED
  question_id: number;
  selected_choice_id: number;
  user_id?: number;         // Optional (derived from JWT)
}

export interface BackendAnswerOut {
  attempt_id: number;
  feedback_id: number;
  is_correct: boolean;
  feedback_text: string;    // NOT "explanation"
  is_attempt_finished: boolean;
  ai_payload: Record<string, unknown>;
}

// ============================================================================
// FRONTEND STATE TYPES (Internal use only)
// ============================================================================

export interface QuizState {
  question: BackendQuestionOut | null;
  selectedChoice: number | null;
  submitted: boolean;
  isCorrect: boolean | null;
  feedbackText: string | null;
  aiPayload: Record<string, unknown> | null;
  isFinished: boolean;
  loading: boolean;
  error: string | null;
  attemptId: number | null;
  questionsAnswered: number;
  correctAnswers: number;
}
